src/github.com/matrix-org/go-neb/services/circleci/circleci.go

```diff
@@ -0,0 +1,276 @@
+	t["repository"] = t["repository_slug"] // Deprecated form but still used everywhere in people's templates
+	t["repository_name"] = p.Reponame
+	t["build_number"] = strconv.Itoa(p.BuildNum)
+	t["build_id"] = t["build_number"] // CircleCI doesn't have a difference between number and ID but to be consistent with TravisCI
+	shaLength := len(p.VcsRevision)
+	if shaLength > 10 {
+		shaLength = 10
+	}
+	t["commit"] = p.VcsRevision[:shaLength] // shortened commit SHA
+	t["author"] = p.CommitterName           // author: commit author name
+	// commit_message: commit message of build
+	// commit_subject: first line of the commit message
+	t["commit_message"] = p.Body
+	subjAndMsg := strings.SplitN(p.Body, "\n", 2)
+	t["commit_subject"] = subjAndMsg[0]
+	if p.Status != "" {
+		t["result"] = p.Status
+	}
+	t["message"] = p.Outcome // message: CircleCI message to the build
+
+	if !p.StartTime.IsZero() && !p.StopTime.IsZero() {
+		t["duration"] = p.StopTime.Sub(p.StartTime).String()
+		t["elapsed_time"] = t["duration"]
+	}
+
+	t["build_url"] = p.BuildURL
+
+	//Map json fields
+	t["vcs_url"] = p.VcsURL
+	t["build_num"] = strconv.Itoa(p.BuildNum)
+	t["branch"] = p.Branch
+	t["vcs_revision"] = p.VcsRevision
+	t["committer_name"] = p.CommitterName
+	t["committer_email"] = p.CommitterEmail
+	t["subject"] = p.Subject
+	t["body"] = p.Body
+	t["why"] = p.Why
+	switch value := p.DontBuild.(type) {
+	case string:
+		t["dont_build"] = value
+	case int:
+		t["dont_build"] = strconv.Itoa(value)
+	case bool:
+		t["dont_build"] = strconv.FormatBool(value)
+	}
+	t["queued_at"] = p.QueuedAt.String()
+	t["start_time"] = p.StartTime.String()
+	t["stop_time"] = p.StopTime.String()
+	t["build_time_millis"] = strconv.Itoa(p.BuildTimeMillis)
+	t["username"] = p.Username
+	t["reponame"] = p.Reponame
+	t["lifecycle"] = p.Lifecycle
+	t["outcome"] = p.Outcome
+	t["status"] = p.Status
+	switch value := p.RetryOf.(type) {
+	case string:
+		t["retry_of"] = value
+	case int:
+		t["retry_of"] = strconv.Itoa(value)
+	}
+	//TODO: Figure out how to map the Steps Slice/Array
+	return t
+}
+
+func outputForTemplate(circleciTmpl string, tmpl map[string]string) (out string) {
+	if circleciTmpl == "" {
+		circleciTmpl = DefaultTemplate
+	}
+	out = circleciTmpl
+	for tmplVar, tmplValue := range tmpl {
+		out = strings.Replace(out, "%{"+tmplVar+"}", tmplValue, -1)
+	}
+	return out
+}
+
+// OnReceiveWebhook receives requests from CircleCI and possibly sends requests to Matrix as a result.
+//
+// If the repository matches a known Git* repository, a notification will be formed from the
+// template for that repository and a notice will be sent to Matrix.
+//
+// Go-NEB cannot register with CircleCI for webhooks automatically. The user must manually add the
+// webhook endpoint URL to their .circleci/config.yml file:
+//    notify:
+//		webhooks:
+//			- url: https://example.com/services/hooks/circle
+//
+// See https://circleci.com/docs/1.0/configuration/#notify for more information.
```

Comment:
note that the circlci 2.0 docs completely omit the webhooks documentation, but based on user reports it seems to still work. Not really sure what they're doing :/